### ISSUE TITLE
Show modern build stats for packages that expose a "exports.browser" build

### ISSUE THREAD

**Issue by GaryLChew** [pastelsky/bundlephobia#391]
**Please describe the feature/suggestion**
With webpack 5, we expect to start seeing package authors offer a modern ["exports.browser" conditional export](https://gist.github.com/sokra/e032a0f17c1721c71cfced6f14516c62#:~:text=browser:%20Compatible%20with%20current%20Spec). This might be the solution to finally [bringing modern javascript to libraries](https://dev.to/garylchew/bringing-modern-javascript-to-libraries-432c) and cutting library bloat by 5-40%.

BundlePhobia is the go-to tool in helping developers avoid library bloat.
It would be great if BundlePhobia could also showcase some information on the modern entry points that these packages offer. This could serve to:
1. Enable application developers to see which libraries offer a modern `"exports.browser"` build.
2. Show off the reduced size of modern builds and encourage library authors and application developers to provide or use the `"exports.browser"` entry point.
3. Prevent package authors from publishing ES5 builds to `"exports.browser"`.

**Describe the solution you'd like**
BundlePhobia's results page could have a "modern"/"non-modern" toggle for packages that offer a `"exports.browser"` conditional export (see below).
![image](https://user-images.githubusercontent.com/30234864/90543985-4e3a8700-e13b-11ea-8ec4-32b8c38e71ce.png)
*(Mock created by @jeremyphilemon)*
If a user clicks on the `"modern"` tab, BundlePhobia will switch from using the top-level `"browser"/"module"/"main"` entry points and instead use the `"exports.browser"` conditional entry point. All sections on the page (i.e. bundle size, download time, history, etc.) will display information for the smaller modern entry point.

In addition, for packages that offer a modern "exports.browser" build, the results page can show a banner along the lines of `This package also ships with modern code in **ES2018**. Save 23% by adding support for exports to your bundler.`

In the case that a package author incorrectly provides a transpiled ES5 build in `"exports.browser"`, the banner can instead a show a warning such as `This package offers a modern "exports.browser" entry point, but does not use any ES2015+ features`. BundlePhobia can implement this syntax detection using detect-es-version (a package I'm in the middle of implementing).

*Implementation Details*
- `bundlephobia.com/result?p=react?exports=browser` will route to the results page for the modern `"exports.browser"` build.
- An exports parameter will be added for all relevant API's (getInfo, getExports, getExportsSizes, getHistory)
- The `exports` parameter will be added as a key in BundlePhobia's Firebase cache.

*Other Notes*
- Implementing this functionality currently requires using webpack's and enhanced-resolve's 5.0.0 beta versions.

**Describe any alternatives you've considered**
N/A

**Comment by GaryLChew** [comment]
@pastelsky @jeremyphilemon
I have a few questions:
1. In our discussions we mentioned showing a badge for packages that offer a modern version. What would we want this to look like?
2. Do we want to show the banner only when the modern tab is selected?
3. To be able to show `"This package also ships with modern code in **ES2018**.` I'm thinking of:
    a. Adding a new '/api/ecma-version?package=${packageString} endpoint to bundlephobia's API
    b. Implementing this API in package-build-stats, adding detect-es-version as a dependency. This getEcmaVersion(packageString) will use the existing InstallationUtils to install the package from the network/cache.
Thoughts on if this approach sounds good? If so, I can start working on a PR.

**Comment by pastelsky** [comment]
> To be able to show "This package also ships with modern code in **ES2018**. I'm thinking of:
a. Adding a new '/api/ecma-version?package=${packageString} endpoint to bundlephobia's API

The `detect-es-version` looks at the AST to find out the ES version. This should be fast? It could also potentially re-use the AST generated by webpack?

If the detection takes a couple of seconds, I don't mind adding this to the main `api/size?package=<>` API instead.

> Do we want to show the banner only when the modern tab is selected
The modern/non-modern tabbing can be taken up as a later step I think. For the first phase, adding a banner on the results page (something like this maybe?)

![image](https://user-images.githubusercontent.com/8946207/91654614-f3544880-eac7-11ea-9869-4921a398a54e.png)

In the medium term, I think we should make an independent badging section below the search box — we seem to be running out of space for more there. cc @jeremyphilemon

> It was suggested that we first "show stats for a modern build then lazy load the non-modern build and have it ready for switch"

 I was actually suggesting doing this the other way around. i.e. show the non-modern build by default and pull up modern numbers if the modern tab is clicked. 

For doing the latter, we'd need to think about how we're going to add this dimension for existing data (package history/composition etc) and how this will be modelled into firebase (separate index? a new property?)

**Comment by GaryLChew** [comment]
> The detect-es-version looks at the AST to find out the ES version. This should be fast? Maybe... Just maybe...It could also potentially re-use the AST generated by webpack?

I've run some benchmarks on the speed [here](https://github.com/GoogleChromeLabs/detect-es-version/blob/master/scripts/results/get-local-package-ecma-version-benchmark-results.md). Re-using webpack's AST is definitely possible (@developit has experimented with this a bit), but I would guess that the overall runtime decrease would only be noticeable for very large packages.

>If the detection takes a couple of seconds, I don't mind adding this to the main api/size?package=<> API instead.
One thing though is that the `This package offers a modern "exports.browser" entry point, but does not use any ES2015+ features` message requires looking at "exports.browser", so we would have to either:
1. Call api/size?package=<> which, despite returning stats for `"main"`/`"module"`/etc., will also return the ecma version for `"exports.browser"`
2. Call api/size?package=<>?exports=browser

>I was actually suggesting doing this the other way around. i.e. show the non-modern build by default and pull up modern numbers if the modern tab is clicked.
Thanks for clarifying, that sounds good too!

**Comment by pastelsky** [comment]
> This package offers a modern "exports.browser" entry point, but does not use any ES2015+ features

Yeah, showing this lint warning for packages is going to be a challenge if we don't build for `exports.browser`. I'm not too keen on having two builds for each listing view right now though. 

This warning seems to be more useful for folks already looking at consuming `exports.browser`. I think this can be added in the second phase when we add support for switching between modern/non-modern bundles and the person is viewing the modern bundle.

**Comment by GaryLChew** [comment]
Ah okay, so you'd just want something like "This package ships with modern code" based on the presence of exports.browser? When do you imagine the 2nd phase being started?

Also, today's my last day at Google, so I think @developit will look into picking this up later.
